napari/_app_model/utils.py

```diff
@@ -0,0 +1,92 @@
+from typing import Union
+
+from app_model.expressions import parse_expression
+from app_model.types import Action, MenuItem, SubmenuItem
+
+from napari._app_model import get_app
+from napari._app_model.constants import MenuGroup, MenuId
+
+MenuOrSubmenu = Union[MenuItem, SubmenuItem]
+
+
+def contains_dummy_action(menu_items: list[MenuOrSubmenu]) -> bool:
+    """True if one of the menu_items is the dummy action, otherwise False.
+
+    Parameters
+    ----------
+    menu_items : list[MenuOrSubmenu]
+        menu items belonging to a given menu
+
+    Returns
+    -------
+    bool
+        True if menu_items contains dummy item otherwise false
+    """
+    for item in menu_items:
+        if hasattr(item, 'command') and 'empty_dummy' in item.command.id:
+            return True
+    return False
+
+
+def is_empty_menu(menu_id: str) -> bool:
+    """Return True if the given menu_id is empty, otherwise False
+
+    Parameters
+    ----------
+    menu_id : str
+        id of the menu to check
+
+    Returns
+    -------
+    bool
+        True if the given menu_id is empty, otherwise False
+    """
+    app = get_app()
+    if menu_id not in app.menus:
+        return True
+    if len(app.menus.get_menu(menu_id)) == 0:
+        return True
+    if len(app.menus.get_menu(menu_id)) == 1 and contains_dummy_action(
+        app.menus.get_menu(menu_id)
+    ):
+        return True
+    return False
+
+
+def no_op() -> None:
+    """Fully qualified no-op to use for dummy actions."""
+
+
+def get_dummy_action(id_key: str, menu_id: MenuId) -> tuple[Action, str]:
+    """Return a dummy action to be used for the given menu.
+
+    id_key will be formatted into the action ID: 'napari.{id_key}.empty_dummy'.
+
+    Parameters
+    ----------
+    id_key: str
+        key to be formatted into the action ID and when expression
+    menu_id: MenuId
+        id of the menu to add the dummy action to
```

Comment:
what is the difference between id_key and menu_id?

Reply:
@jni the `id_key` will be used to form the id of the action object, while the `menu_id` is the menu the dummy action will be placed in

Reply:
the `id_key` is also used to form a unique context key for the action

Reply:
... But isn't the id_key the same as the menu_id in general? Or is this because it's the menu_id key stripped of the prefixes? (in which case, maybe we should just do that in the function?)

**update** yes it is, [here](https://github.com/napari/napari/pull/7038/files#diff-cf45dd214c7efa10353f0da02c807329f8b07d4115667787095eff0d36c65079R120):

```python
        id_key = menu_id.split('/')[-1]
        dummmy_action, context_key = get_dummy_action(id_key, menu_id)
```